serialization/numpy_npy.ksy

```diff
@@ -0,0 +1,232 @@
+          "descr" : dtype.descr
+              An object that can be passed as an argument to the numpy.dtype() constructor to create the array's dtype.
+          "fortran_order" : bool
+              Whether the array data is Fortran-contiguous or not.
+              Since Fortran-contiguous arrays are a common form of non-C-contiguity, we allow them to be written directly to disk for efficiency.
+          "shape" : tuple of int
+              The shape of the array.
+      For repeatability and readability, this dictionary is formatted using pprint.pformat() so the keys are in alphabetic order. A writer SHOULD implement this if possible. A reader MUST NOT depend on this.
+  - id: data_initial
+    size-eos: true
+    type: ndarray(-1)
+    doc: |
+      Following the header comes the array data. If the dtype contains Python objects (i.e. dtype.hasobject is True), then the data is a Python pickle of the array. Otherwise the data is the contiguous (either C- or Fortran-, depending on fortran_order) bytes of the array. Consumers can figure out the number of bytes by multiplying the number of elements given by the shape (noting that shape=() means there is 1 element) by dtype.itemsize.
+instances:
+  data:
+    value: data_initial.data
+  dims_m_2:
+    value: parsed_header.dimensions.as<u8> - 2
+  parsed_header:
+    pos: 0
+    type: numpy_parsed_header(header)
+    doc: |
+      # save it as numpy_parsed_header.py
+      import json
+      from kaitaistruct import KaitaiStruct
+      import re
+      
+      braceFixingRx = re.compile(",\s*([\]\}])")
+      
+      def literalEvalSurrogate(s:str):
+        #A surrogate of `ast.literal_eval` as a solution for the langs not having python syntax parser
+        #BUG in KSC: `doc` cannot contain triple braces
+        s = s.strip()
+        s = s.replace("(", "[").replace(")", "]")
+        s = braceFixingRx.subn(lambda x: x.group(1), s)[0]
+        s = s.replace("True", "true").replace("False", "false")
+        s = s.replace("'", '"'))
+        return json.loads(s)
+      
+      class NumpyParsedHeader(KaitaiStruct):
+        __slots__ = ("item_type", "fortran_order", "shape", "dimensions")
+        def __init__(self, header_str: str, _io, _parent=None, _root=None):
+          self._parent = _parent
+          header = literalEvalSurrogate(header_str)
+          self.item_type = header["descr"]
+          if not isinstance(self.item_type , str):
+            raise NotImplementedError(repr(self.item_type))
+          self.fortran_order = header["fortran_order"]
+          self.shape = header["shape"]
+          self.dimensions = len(self.shape)
+types:
+  version:
+    seq:
+      - id: major
+        type: u1
+      - id: minor
+        type: u1
+  array:
+    params:
+      - id: idx
+        type: u8
+    seq:
+      - id: data
+        doc: "S<n> for n > 1 is currently out of scope"
+        type:
+          switch-on: _root.parsed_header.item_type.as<str>
```

Comment:
On the topic of dtypes, while reading the numpy docs, I noticed that numpy supports [many more complex dtypes](https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html#specifying-and-constructing-data-types) than just single scalars. For these dtypes, `"descr"` in the header dictionary is a list or dictionary rather than a string.

For example, calling `numpy.save` with this array:

```python
numpy.array([(12, 3.4), (56, 7.8)], dtype=[("", "u2"), ("", "f8")])
```

produces a file with this header (wrapped for readability):

```python
{
    'descr': [('foo', '<u2'), ('bar', '<f8')],
    'fortran_order': False,
    'shape': (2,),
}
```

I don't think there's any way to support arbitrary complex dtypes in this spec, so we should probably stick to simple dtypes only and document this limitation.

Reply:
Yes, they are currently out of scope.